CPDShell/Core/algorithms/knn_algorithm.py

```diff
@@ -0,0 +1,175 @@
+        """Finds coordinates of change points (localizes them) in window.
+
+        :param window: part of global data for finding change points.
+        :return: list of window change points.
+        """
+        self.__process_data(window)
+        return self.__change_points.copy()
+
+    def __process_data(self, window: Iterable[float | np.float64]) -> None:
+        """
+        Processes a window of data to detect/localize all change points depending on working mode.
+
+        :param window: part of global data for change points analysis.
+        """
+        sample = deque(window)
+        sample_size = len(sample)
+        if sample_size == 0:
+            return
+
+        # Preparing.
+        self.__change_points: list[int] = []
+        self.__change_points_count = 0
+
+        # Building graph.
+        self.__knngraph = knngraph.KNNGraph(window, self.__metric, self.__k)
+        self.__knngraph.build()
+
+        # Examining each point.
+        # Boundaries are always change points.
+        first_point = int(len(window) / 4)
+        last_point = int(len(window) * 0.75)
+
+        for time in range(first_point, last_point):
+            statistics = self.__calculate_statistics_in_point(time, len(window))
+            # print(time, statistics)
+            if self.__check_change_point(statistics):
+                self.__change_points.append(time)
+                self.__change_points_count += 1
+
+    def __calculate_statistics_in_point(self, time: int, window_size: int) -> float:
+        """
+        Calculate the statistics of the KNN graph in specified point.
+
+        :param time: index of point in the given sample to calculate statistics relative to it.
+        :param window_size: size of sample to analyze.
+        """
+        assert self.__knngraph is not None, "Graph should not be None."
+
+        k = self.__k
+        n = window_size
+        n_1 = time
+        n_2 = n - time
+
+        if n <= k:
+            # Unable to analyze sample due to its size.
+            return -k
+
+        h = 4 * (n_1 - 1) * (n_2 - 1) / ((n - 2) * (n - 3))
+
+        sum_1 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(i, j) * self.__knngraph.check_for_neighbourhood(j, i)
+            for i in range(window_size)
+            for j in range(window_size)
+        )
+
+        sum_2 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(j, i) * self.__knngraph.check_for_neighbourhood(m, i)
+            for i in range(window_size)
+            for j in range(window_size)
+            for m in range(window_size)
+        )
+
+        expectation = 4 * k * n_1 * (n_2) / (n - 1)
+        variance = (expectation / k) * (h * (sum_1 + k - (2 * k**2 / (n - 1))) + (1 - h) * (sum_2 - k**2))
+        deviation = sqrt(variance)
+
+        permutation: np.array = np.arange(window_size)
+        # np.random.shuffle(permutation) # It seems that random permutation spoils the result
```

Comment:
It seems to be unrelated for the current code, but looks like an interesting research question, if I understand correctly.

Reply:
I decided to make a small research on this topic. It seems that it spoils result on some good distributions and it can make it more precise on bad distributions (such as weibull).